app/org/sagebionetworks/bridge/dynamodb/DynamoUserConsentDao.java

```diff
@@ -6,81 +6,127 @@
+        DynamoUserConsent3 consent3 = getUserConsent3(healthCode, studyConsent.getStudyKey());
+        if (consent3 == null) {
+            consent3 = new DynamoUserConsent3(healthCode, studyConsent.getStudyKey());
+            consent3.setConsentCreatedOn(studyConsent.getCreatedOn());
+            consent3.setSignedOn(signedOn);
+            mapperV3.save(consent3);
+        }
+        if (consent2 == null) {
+            try {
+                consent2 = new DynamoUserConsent2(healthCode, studyConsent.getStudyKey());
+                consent2.setConsentCreatedOn(studyConsent.getCreatedOn());
+                consent2.setSignedOn(signedOn);
+                mapperV2.save(consent2);
+            } catch(Exception e) {
+                // compensate and then rethrow exception
+                mapperV3.delete(consent3);
+                throw e;
             }
-            consent.setSignedOn(DateTime.now(DateTimeZone.UTC).getMillis());
-            mapper.save(consent);
-        } catch (ConditionalCheckFailedException e) {
-            throw new EntityAlreadyExistsException(consent);
         }
-        return consent;
+        return consent2;
     }
 
     @Override
     public boolean withdrawConsent(String healthCode, StudyIdentifier studyIdentifier) {
-        DynamoUserConsent2 consent = (DynamoUserConsent2) getUserConsent(healthCode, studyIdentifier);
-        if (consent != null) {
-            mapper.delete(consent);
-            return true;
+        // In first step of migration, if user withdraws consent, we delete UserConsent3 record. Once all records are
+        // synchronized and we're ready to switch over, they will be marked with a withdrawal timestamp and queries 
+        // will be adjusted accordingly. We will also need to coordinate this timestamp with the historical signature 
+        // record in Stormpath.
+        DynamoUserConsent2 consent2 = getUserConsent2(healthCode, studyIdentifier.getIdentifier());
+        DynamoUserConsent3 consent3 = getUserConsent3(healthCode, studyIdentifier.getIdentifier());
+        
+        if (consent2 == null && consent3 == null) {
+            return false;
         }
-        return false;
+        // delete 3 then delete 2, re-save 3 if 2 fails
+        if (consent3 != null) {
+            mapperV3.delete(consent3);
+        }
+        if (consent2 != null) {
+            try {
+                mapperV2.delete(consent2);    
+            } catch(Exception e) {
+                // compensate for delete failure. Need to vacate the version attribute
+                consent3.setVersion(null);
+                mapperV3.save(consent3);
+                throw e;
+            }
+        }
+        return true;
     }
 
     @Override
     public boolean hasConsented(String healthCode, StudyIdentifier studyIdentifier) {
-        return getUserConsent(healthCode, studyIdentifier) != null;
+        return getUserConsent2(healthCode, studyIdentifier.getIdentifier()) != null;
     }
```

Comment:
As per previous comment, this should check getUserConsent3() != null, then fall back to getUserConsent2() != null. Alternatively, move the fallback logic to getUserConsent(), so that getUserConsent() returns the "true" consent.